.github/workflows/ci.yml

```diff
@@ -0,0 +1,240 @@
+defaults:
+  run:
+    shell: bash
+
+jobs:
+  build:
+    # Build an F* binary package: a fully-bootstrapped stage 2 compiler,
+    # with its plugins, a fully checked library (i.e. with .checked)
+    # files and compiled versions of fstar_lib and fstar_plugin_lib.
+    # Also, we build a full stage 1 compiler and its libraries, and
+    # make all possible artifacts.
+    runs-on: ubuntu-latest
+    container: mtzguido/dev-base
+    steps:
+      - name: Cleanup
+        run: find . -delete
+      - run: echo "HOME=/home/user" >> $GITHUB_ENV
+      - uses: mtzguido/set-opam-env@master
+
+      - name: Checkout
+        uses: actions/checkout@master
+        with:
+          path: FStar
+
+      - name: Produce all artifacts
+        run: make -skj$(nproc) all-packages FSTAR_VERSION=ci ADMIT=1
+        working-directory: FStar
+
+      # Upload the archives.
+      - uses: actions/upload-artifact@v4
+        with:
+          path: FStar/fstar-ci.tar.gz
+          name: fstar-ci.tar.gz
+          retention-days: 3
+      - uses: actions/upload-artifact@v4
+        with:
+          path: FStar/fstar-ci-stage1.tar.gz
+          name: fstar-ci-stage1.tar.gz
+          retention-days: 3
+      - uses: actions/upload-artifact@v4
+        with:
+          path: FStar/fstar-ci-src.tar.gz
+          name: fstar-ci-src.tar.gz
+          retention-days: 3
+      - uses: actions/upload-artifact@v4
+        with:
+          path: FStar/fstar-ci-stage1-src.tar.gz
+          name: fstar-ci-stage1-src.tar.gz
+          retention-days: 3
+
+      # Upload full repo too, for Pulse. Note: we explicitly run
+      # 'make 2' at this point to generate the out/ directory, as the previous
+      # targets do not. Also, remove the previous archives so they don't
+      # blow up the size of this artifact.
+      - run: rm -f FStar/fstar-ci*.tar.gz
+      - run: make 2
+        working-directory: FStar
+
+      - uses: mtzguido/gci-upload@master
+        with:
+          name: fstar-repo
+          path: FStar
+          extra: --exclude=./stage*/dune/_build
+          hometag: FSTAR
+
+      # FIXME: Ideally, we could upload the artifacts as soon as each of
+      # them is created, and start the subsequent jobs at that instant too.
+      # Is that even doable...?
+
+  check-friends:
+    needs: build
+    name: friends
+    uses: ./.github/workflows/check-friends.yml
+
+  # This checks that the stage3 extracted fstarc matches exactly with stage2.
+  # There is some overlap with the job above, but the testing jobs do not
+  # need this bit at all and we have plenty of parallelism for now... so just
+  # spawn another one.
+  check-stage3-diff:
+    runs-on: ubuntu-latest
+    container: mtzguido/dev-base
+    steps:
+      - name: Cleanup
+        run: find . -delete
+      - run: echo "HOME=/home/user" >> $GITHUB_ENV
+      - uses: mtzguido/set-opam-env@master
+
+      - name: Checkout
+        uses: actions/checkout@master
+
+      - name: Check for a stage 3 diff
```

Comment:
Should we append this step to build to reduce CI usage?

Reply:
I was wondering how best to split all this.. doing that does save on total usage, but splitting it away reduces the time until the test jobs can start. (Testing could also be done in this same job, which would maximize parallelism with stage3, but I liked the idea of a starting from a clean slate, and allows subprojects to be checked even if some test fails.) Another bonus is that if the stage3 diff fails for some reason, we still go ahead with normal tests.

Maybe we can just record the state of the repo in an artifact (in fact it's done already, for Pulse) and start the stage3 check from there... wdyt?